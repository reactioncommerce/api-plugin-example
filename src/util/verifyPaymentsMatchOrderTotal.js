import accounting from "accounting-js";
import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";

/**
 * @summary Given an array of payment input and an order total,
 *   checks that the sum of all payment amounts matches the order total.
 *   Throws a ReactionError if not.
 * @param {Object[]} paymentsInput Array of PaymentInput objects, potentially empty
 * @param {Number} orderTotal The grand total of the order these payments are for.
 * @returns {undefined}
 */
export default function verifyPaymentsMatchOrderTotal(paymentsInput, orderTotal) {
  const paymentTotal = paymentsInput.reduce((sum, paymentInput) => sum + paymentInput.amount, 0);

  // In order to prevent mismatch due to rounding, we convert these to strings before comparing. What we really
  // care about is, do these match to the specificity that the shopper will see (i.e. to the scale of the currency)?
  // No currencies have greater than 3 decimal places, so we'll use 3.
  const paymentTotalString = accounting.toFixed(paymentTotal, 3);
  const orderTotalString = accounting.toFixed(orderTotal, 3);

  if (paymentTotalString !== orderTotalString) {
    Logger.debug("Error creating payments for a new order. " +
      `Order total (${orderTotalString}) does not match total of all payment amounts (${paymentTotalString}).`);
    throw new ReactionError("payment-failed", "Total of all payments must equal order total");
  }
}
