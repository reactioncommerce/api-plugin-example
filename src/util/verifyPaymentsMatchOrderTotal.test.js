import verifyPaymentsMatchOrderTotal from "./verifyPaymentsMatchOrderTotal.js";

test("throws if does not match", () => {
  const payments = [
    { amount: 5 }
  ];

  expect(() => verifyPaymentsMatchOrderTotal(payments, 6)).toThrowErrorMatchingSnapshot();
});

test("does not throw if matches", () => {
  const payments = [
    { amount: 5 }
  ];

  expect(() => verifyPaymentsMatchOrderTotal(payments, 5)).not.toThrow();
});

test("is not confused by JavaScript floating point math errors", () => {
  // Do 9.99+50+40 in JS console. Result is 99.99000000000001 due to inaccuracy
  // of JS floating point math.
  const payments = [
    { amount: 9.99 },
    { amount: 50 },
    { amount: 40 }
  ];

  expect(() => verifyPaymentsMatchOrderTotal(payments, 99.99)).not.toThrow();
});
