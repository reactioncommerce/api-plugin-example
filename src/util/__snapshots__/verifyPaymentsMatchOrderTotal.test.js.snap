// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if does not match 1`] = `"Total of all payments must equal order total"`;
